File: Misc/aho-corasick.py
#!/usr/bin/env python3

from collections import deque

# Aho-Corasick, Efficient String Matching, 1975
# using dictionnaries (fast but memory expensive)
class ACTrie:
    def __init__(self,W):
        self.W = W[:]   # copy for safety
        self.G = [{}]   # transitions
        self.O = [[]]   # final states
        self.F = [None] # fail function
        self.B = [None] # dict backlink
        for iw in range(len(self.W)):
            w = self.W[iw]
            s = 0
            for c in w:
                if c in self.G[s]:
                    s = self.G[s][c]
                else:
                    s0 = self.new_state()
                    self.G[s][c] = s0
                    s = s0
            self.O[s].append(iw)
        # failure computation (BFS)
        Q = deque([0])
        while Q:
            s = Q.popleft()
            for c in self.G[s]:
                s0 = self.G[s][c]
                Q.append(s0)
                if s==0: # s0 at depth 1
                    self.F[s0] = 0
                else:
                    f = self.F[s]
                    while self.g(f,c)==None:
                        f = self.F[f]
                    f0 = self.g(f,c)
                    self.F[s0] = f0
                    # IMPLEM. 1 - on complete la sortie de s0
                    #self.O[s0] += self.O[f0]
                    # IMPLEM. 2 - backlink
                    self.B[s0] = f0 if self.O[f0] else self.B[f0]
            
    def g(self,s,c): # transition function completed for 0 -*-> 0
        if c in self.G[s]:
            return self.G[s][c]
        return 0 if s==0 else None

    def new_state(self):
        s = len(self.G)
        self.G.append({})
        self.O.append([])
        self.F.append(None)
        self.B.append(None)
        return s

    def find(self,S):
        s = 0
        for i in range(len(S)):
            c = S[i]
            while self.g(s,c)==None:
                s = self.F[s]
            s = self.g(s,c)
            # IMPLEM. 1 dans laquelle on complete l'output
            # if self.O[s]: # motif(s) trouve(s)
                # renvoie les indices des motifs
                # yield (i,self.O[s])
                # renvoie les motifs
                # yield (i,list(map(lambda iw: self.W[iw], self.O[s])))
            # IMPLEM. 2 dans laquelle on utilise des backlinks
            if self.O[s] or self.B[s]: # motif direct ou backlink
                # renvoie le noeud du trie
                yield (i,s)

    # IMPLEM. 2 - calculer l'output d'un noeud
    # (version iterative, parce que Python)
    def output(self,s):
        while s:
            yield from self.O[s]
            s = self.B[s]

A = ACTrie(['he','she','is','hers'])
for (i,s) in A.find('ishers'):
    print(i,[A.W[a] for a in A.output(s)])

B = ACTrie(['aaa','a','aa'])
for (i,s) in B.find('aaaaa'):
    print(i,[B.W[a] for a in B.output(s)])


# =============================================
# Aho-Corasick, Efficient String Matching, 1975
# version with fixed alphabet size
# using lists instead of dictionnaries
# (slighly slower but potentially less memory
#  exensive, depending on the cases however)
Alpha = 26 # alphabet size

def num(c): # char to num
    return ord(c)-ord('a')

class ACTrie:
    def __init__(self,W):
        self.W = W
        self.G = [[None]*Alpha] # transitions
        self.O = [[]]           # final states
        self.F = [None]         # fail function
        self.B = [None]         # dict backlink
        for iw in range(len(self.W)):
            w = self.W[iw]
            s = 0
            for c in w:
                c = num(c)
                if self.G[s][c]!=None:
                    s = self.G[s][c]
                else:
                    s0 = self.new_state()
                    self.G[s][c] = s0
                    s = s0
            self.O[s].append(iw)
        # failure computation (BFS)
        Q = deque([0])
        while Q:
            s = Q.popleft()
            for c in range(Alpha):
                s0 = self.G[s][c]
                if s0==None:
                    continue
                Q.append(s0)
                if s==0: # s0 at depth 1
                    self.F[s0] = 0
                else:
                    f = self.F[s]
                    while self.g(f,c)==None:
                        f = self.F[f]
                    f0 = self.g(f,c)
                    self.F[s0] = f0
                    self.B[s0] = f0 if self.O[f0] else self.B[f0]
            
    def g(self,s,c): # transition function completed for 0 -*-> 0
        if self.G[s][c]!=None:
            return self.G[s][c]
        return 0 if s==0 else None

    def new_state(self):
        s = len(self.G)
        self.G.append([None]*Alpha)
        self.O.append([])
        self.F.append(None)
        self.B.append(None)
        return s

    def find(self,S):
        s = 0
        for i in range(len(S)):
            c = num(S[i])
            while self.g(s,c)==None:
                s = self.F[s]
            s = self.g(s,c)
            if self.O[s] or self.B[s]: # motif(s) trouve(s)
                yield (i,s)
    
    def output(self,s):
        while s:
            yield from self.O[s]
            s = self.B[s]


A = ACTrie(['he','she','is','hers'])
for (i,s) in A.find('ishers'):
    print(i,[A.W[a] for a in A.output(s)])

B = ACTrie(['aaa','a','aa'])
for (i,s) in B.find('aaaaa'):
    print(i,[B.W[a] for a in B.output(s)])

File: Misc/aho_corasick.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

/*
  Aho-Corasick, Efficient String Matching, 1975
*/

const int Alpha = 26;

int num(char c) {
  return (int)c-(int)'a';
}

typedef pair<int,int> match;

struct ACTrie {
  vector<string> W;
  vector< vector<int> > G,O;
  vector<int> F,B;

  int new_state();
  int g(int s, int c) const;
  vector<match> find(const string &S) const;
  vector<int> output(int s);
  
  ACTrie(const vector<string> &words) {
    W = words; // copy
    G.push_back(vector<int>(Alpha,-1));
    O.resize(1);
    F.push_back(-1);
    B.push_back(-1);
    for (int iw=0; iw<(int)W.size(); ++iw) {
      string w = W[iw];
      int s = 0;
      for (int ic=0; ic<(int)w.size(); ++ic) {
	int c = num(w[ic]);
	if (G[s][c]>=0) s = G[s][c];
	else {
	  int s0 = new_state();
	  G[s][c] = s0;
	  s = s0;
	}
      }
      O[s].push_back(iw);
    }
    // failure computation (BFS)
    queue<int> Q;
    Q.push(0);
    while (!Q.empty()) {
      int s = Q.front();
      Q.pop();
      for (int c=0; c<Alpha; ++c) {
	int s0 = G[s][c];
	if (s0>0) {
	  Q.push(s0);
	  if (s==0) F[s0] = 0;
	  else {
	    int f = F[s];
	    while (g(f,c)<0) f = F[f];
	    int f0 = g(f,c);
	    F[s0] = f0;
	    B[s0] = O[f0].empty() ? B[f0] : f0;
	  }
	}
      }
    }
  }
};

int ACTrie::new_state() {
  int s = G.size();
  G.push_back(vector<int>(Alpha,-1));
  O.resize(O.size()+1);
  F.push_back(-1);
  B.push_back(-1);
  return s;
}

int ACTrie::g(int s, int c) const {
  if (G[s][c]>=0) return G[s][c];
  return s==0 ? 0 : -1;
}

vector<match> ACTrie::find(const string &S) const {
  vector<match> M;
  int s = 0;
  for (int i=0; i<(int)S.size(); ++i) {
    int c = num(S[i]);
    while (g(s,c)<0) s = F[s];
    s = g(s,c);
    // renvoie le noeud du trie
    if (!O[s].empty() || B[s]>0) M.push_back(make_pair(i,s));
  }
  return M;
}

vector<int> ACTrie::output(int s) {
  vector<int> R;
  while (s>0) {
    for (vector<int>::iterator it=O[s].begin(); it!=O[s].end(); ++it)
      R.push_back(*it);
    s = B[s];
  }
  return R;
}

int main() {
  vector<string> W = {"he","she","is","hers"};
  ACTrie A(W);
  vector<match> M = A.find("ishers");
  for (vector<match>::iterator it=M.begin(); it!=M.end(); ++it) {
    vector<int> O = A.output(it->second);
    for (vector<int>::iterator jt=O.begin(); jt!=O.end(); ++jt)
      cout << '"' << W[*jt] << "\" found at " << it->first << endl; 
  }
  return 0;
}
